crypto/src/base58.rs

```diff
@@ -60,25 +56,19 @@ pub trait FromBase58Check {
 
 impl ToBase58Check for [u8] {
     fn to_base58check(&self) -> Result<String, ToBase58CheckError> {
-        if self.len() > 128 {
-            return Err(ToBase58CheckError::DataTooLong);
-        }
         // 4 bytes checksum
         let mut payload = Vec::with_capacity(self.len() + 4);
         payload.extend(self);
         let checksum = double_sha256(self);
         payload.extend(&checksum[..4]);
 
-        Ok(payload.to_base58())
+        Ok(bs58::encode(payload).into_string())
     }
 }
 
 impl FromBase58Check for str {
     fn from_base58check(&self) -> Result<Vec<u8>, FromBase58CheckError> {
-        if self.len() > 128 {
-            return Err(FromBase58CheckError::DataTooLong);
-        }
-        match self.from_base58() {
+        match bs58::decode(self).into_vec() {
             Ok(payload) => {
                 if payload.len() >= Self::CHECKSUM_BYTE_SIZE {
                     let data_len = payload.len() - Self::CHECKSUM_BYTE_SIZE;
```

Comment:
Perhaps we can avoid a redundant allocation here?

```rust
        match bs58::decode(self).into_vec() {
            Ok(mut payload) => {
                if payload.len() >= Self::CHECKSUM_BYTE_SIZE {
                    let data_len = payload.len() - Self::CHECKSUM_BYTE_SIZE;
                    let data = &payload[..data_len];
                    let checksum_provided = &payload[data_len..];

                    let checksum_expected = double_sha256(data);
                    let checksum_expected = &checksum_expected[..4];

                    if checksum_expected == checksum_provided {
                        // here, instead of allocating a new Vec, reuse payload with checksum stripped
                        payload.truncate(data_len);
                        Ok(payload)
                    } else {
                        Err(FromBase58CheckError::InvalidChecksum)
                    }
                } else {
                    Err(FromBase58CheckError::MissingChecksum)
                }
            }
            Err(_) => Err(FromBase58CheckError::InvalidBase58),
        }
```

Reply:
great spot! thanks, have done